src/clusterfuzz/_internal/bot/tasks/utasks/uworker_io.py

```diff
@@ -0,0 +1,245 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Module for dealing with input and output (I/O) to a uworker."""
+
+import base64
+import datetime
+import json
+import tempfile
+import uuid
+
+from google.cloud import ndb
+
+from clusterfuzz._internal.datastore import data_types
+from clusterfuzz._internal.google_cloud_utils import storage
+from clusterfuzz._internal.metrics import logs
+
+
+def generate_new_io_file_name():
+  """Generates a new I/O file name."""
+  return str(uuid.uuid4()).lower()
+
+
+def get_uworker_io_gcs_path():
+  """Returns a GCS path for uworker I/O."""
+  # Inspired by blobs.write_blob.
+  io_bucket = storage.uworker_io_bucket()
+  io_file_name = generate_new_io_file_name()
+  if storage.get(storage.get_cloud_storage_file_path(io_bucket, io_file_name)):
+    raise RuntimeError(f'UUID collision found: {io_file_name}.')
+  return f'/{io_bucket}/{io_file_name}'
+
+
+def get_uworker_output_urls():
+  """Returns a signed download URL for the uworker to upload the output and a
+  GCS url for the tworker to download the output."""
+  gcs_path = get_uworker_io_gcs_path()
+  # Note that the signed upload URL can't be directly downloaded from.
+  return storage.get_signed_upload_url(gcs_path), gcs_path
+
+
+def get_uworker_input_urls():
+  """Returns a signed download URL for the uworker to download the input and a
+  GCS url for the tworker to upload it (this happens first)."""
+  gcs_path = get_uworker_io_gcs_path()
+  return storage.get_signed_download_url(gcs_path), gcs_path
+
+
+def upload_uworker_input(uworker_input):
+  """Uploads input for the untrusted portion of a task."""
+  signed_download_url, gcs_path = get_uworker_input_urls()
+
+  with tempfile.NamedTemporaryFile() as uworker_input_file:
+    with open(uworker_input_file.name, 'w') as fp:
+      fp.write(uworker_input)
+    if not storage.copy_file_to(uworker_input_file.name, gcs_path):
+      raise RuntimeError('Failed to upload uworker_input.')
+  return signed_download_url
+
+
+def make_ndb_entity_input_obj_serializable(obj):
+  """Returns a dictionary that can be JSON serialized representing an NDB
+  entity. Does not include datetime fields."""
+  obj_dict = obj.to_dict()
+  # TOOD(metzman): Handle datetimes.
+  for key in list(obj_dict.keys()):
```

Comment:
Do we need to cast it into a list?
I thought `for key in obj_dict` is sufficient to iterate keys.